utilities/ideascale-importer/ideascale_importer/snapshot_importer/importer.py

```diff
@@ -94,51 +106,97 @@ def __init__(
+        conn = await ideascale_importer.db.connect(self.database_url)
+
+        # Check if a final snapshot already exists
+        row = await conn.fetchrow("SELECT final FROM snapshot WHERE event = $1", self.event_id)
+        if row is not None and row["final"]:
+            raise FinalSnapshotAlreadyPresent()
+
+        await conn.close()
+
     async def _fetch_parameters(self):
+        # Fetch event parameters
+        conn = await ideascale_importer.db.connect(self.database_url)
+
+        row = await conn.fetchrow(
+            "SELECT registration_snapshot_time, snapshot_start, voting_power_threshold, max_voting_power_pct FROM event WHERE row_id = $1",
+            self.event_id,
+        )
+        if row is None:
+            raise FetchParametersFailed("Failed to get event parameters from the database: " f"event_id={self.event_id} not found")
+
+        self.voting_power_cap = row["max_voting_power_pct"]
+        self.min_stake_threshold = row["voting_power_threshold"]
+        self.snapshot_start_time = row["snapshot_start"]
+        self.registration_snapshot_time = row["registration_snapshot_time"]
+
+        if self.snapshot_start_time is None or self.registration_snapshot_time is None:
+            raise FetchParametersFailed(
+                "Missing snapshot timestamps for event in the database:"
+                f" snapshot_start={None if self.snapshot_start_time is None else self.snapshot_start_time.isoformat()}"
+                f" registration_snapshot_time={None if self.registration_snapshot_time is None else self.registration_snapshot_time.isoformat()}"
+            )
+
+        logger.info(
+            "Got event parameters",
+            min_stake_threshold=self.min_stake_threshold,
+            voting_power_cap=float(self.voting_power_cap),
+            snapshot_start=None if self.snapshot_start_time is None else self.snapshot_start_time.isoformat(),
+            registration_snapshot_time=None
+            if self.registration_snapshot_time is None
+            else self.registration_snapshot_time.isoformat(),
+        )
+
+        await conn.close()
+
         if not self.skip_snapshot_tool_execution:
             # Fetch max slot
-            logger.info("Querying max slot parameter")
             conn = await ideascale_importer.db.connect(
                 f"postgres://{self.config.dbsync_database.user}:"
                 + f"{self.config.dbsync_database.password}@{self.config.dbsync_database.host}"
                 f"/{self.config.dbsync_database.db}"
             )
 
+            # Fetch slot number and time from the block right before or equal the registration snapshot time
             row = await conn.fetchrow(
-                "SELECT slot_no FROM block WHERE time <= $1 ORDER BY time DESC LIMIT 1", self.config.snapshot_tool.max_time
+                "SELECT slot_no, time FROM block WHERE time <= $1 AND slot_no IS NOT NULL ORDER BY slot_no DESC LIMIT 1",
+                self.registration_snapshot_time,
```

Comment:
This should always return the latest block right before registration shapshot time